Notifications form

keithjgrant opened the pull request:
##### SUMMARY
Adds notification template form and add/edit pages

addresses #7876 

Note: this probably needs a touch more polish but I wanted to get some eyes on it while I'm out on PTO next week

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

keithjgrant committed:
flush out notification form and type subform

keithjgrant committed:
fix initial values for notification type fields

keithjgrant committed:
add custom notification messages subform

keithjgrant committed:
reset notification messages to defaults when switching to new type

keithjgrant committed:
only submit customized notification messages if changed

keithjgrant committed:
create ArrayTextField component

keithjgrant committed:
add some notification form tests; notification add screen

AlexSCorey commented:
@keithjgrant 
Just poking at this a bit:
1) I was able to create an email notification, but then I edited it, and removed the recipients list (which says it's required) and I was able to still save.  However, it did not save the removed, recipients list
2) On the list, it seems like in most places where we have an inline label we are not using a colon to separate the label from the value.  We do use a colon in Survey List, but that's the only place that I found it in a brief search.  @trahman73 can you comment on whether to use the colon?
3) Disable the save button.  This is a new pattern that we need the implement here and in all other forms. (https://github.com/ansible/awx/issues/7773)

trahman73 commented:
@AlexSCorey That's correct, we do not use a colon to separate a label from it's value.

mabashian commented on awx/ui_next/src/screens/NotificationTemplate/shared/CustomMessagesSubForm.jsx:
```diff
@@ -0,0 +1,232 @@
+            css="margin-bottom: var(--pf-c-content--MarginBottom)"
+          >
+            <small>
+              Use custom messages to change the content of notifications sent
```
Hrmmm we'll have to figure out how to translate all of this

mabashian committed:
Fix broken credential form tests after id change

mabashian commented:
@keithjgrant I went ahead and rebased this to make it a bit easier to review.  I noticed that there were some failing tests in the cred form test suite so I pushed a fix for that.  It looks like the only thing left that's failing in the awx-ui-next is a lint failure:

```
2020-08-26 14:15:03.399065 | static | /awx_devel/awx/ui_next/src/components/FormField/ArrayTextField.jsx
2020-08-26 14:15:03.399143 | static |   43:19  error  'value' is already declared in the upper scope  no-shadow
```

mabashian commented:
This may be unrelated to your PR but after I save a change to a notification and get redirected to the details view, I don't see that change immediately reflected.  I have to refresh the page.  Perhaps there's some kind of race going on?

![notif_details](https://user-images.githubusercontent.com/9889020/91317234-c9c5b500-e787-11ea-8c08-ae2e78f447d4.gif)

keithjgrant commented:
> This may be unrelated to your PR but after I save a change to a notification and get redirected to the details view, I don't see that change immediately reflected. I have to refresh the page. Perhaps there's some kind of race going on?

I remember this being an issue with other details pages a while back. I'll track down what we did as a fix then update this one to the same behavior

keithjgrant committed:
ensure notification template re-loads after saving

keithjgrant committed:
fix lint error

keithjgrant committed:
mark text for translation

keithjgrant commented:
recheck

keithjgrant committed:
trying to force tests to re-run

keithjgrant committed:
fix new test from rebase

tiagodread commented:
- Trying to create  a notification without type organization and notification type (only name) The UI doesn’t show any error message

- Trying to create a notification without organization the message "This field is required." is showing twice:

> 1. Fill name
> 2. Choose notification type `Grafana`
> 3. Fill Grafana URL and Grafana API Key
> 4. Click on Save button

![image](https://user-images.githubusercontent.com/12464858/92179545-422d1580-ee1b-11ea-9106-ed4082d18dae.png)

- After saving a notification on the details tab the metadata `Destination Channels` shows all destinations without any space and this happens on all the places where you can specify multiple channels or emails to receive a notification:

![image](https://user-images.githubusercontent.com/12464858/92179889-1199ab80-ee1c-11ea-816a-ea514bf7ad5b.png)

- Trying to create a notification I have to check and uncheck the `Disable SSL verification` if i want to save the form without checking this option (this also happens in all forms that use this select field):

![notfcation_0](https://user-images.githubusercontent.com/12464858/92180104-9ab0e280-ee1c-11ea-8488-fcf7bb850d76.gif)

tiagodread commented:
- Trying to create a notification type mattermost the field `Icon URL`must is required even the label showing not required(this also happens for rocketchat):

![notfcation_1](https://user-images.githubusercontent.com/12464858/92180767-3abb3b80-ee1e-11ea-889f-83657d2e9688.gif)

- I don’t know why, but when you click on any label on `pagerduty` form 2x the page crashes totally and this just happens on this form (I verified all of them):

![notfcation_2](https://user-images.githubusercontent.com/12464858/92181536-d6997700-ee1f-11ea-8a4a-4745a8d5f057.gif)

Console:

`react-dom.development.js:13413 Uncaught Error: Objects are not valid as a React child (found: object with keys {_activeMessages, _catalogs, t, select, plural, selectOrdinal, _dev, _activeLanguageData, _language, _locales, _missing}). If you meant to render a collection of children, use an array instead.`

- Trying to create a notification type twilio the fields `Source phone number` and `Destination SMS number(s)` are accepting letters, is this correct ?

![image](https://user-images.githubusercontent.com/12464858/92181807-8969d500-ee20-11ea-80e4-48971cff2047.png)

- Trying to create a email notification type, the fields `Host, Recipient list, Port and Time out` doesn’t have validation, so I can save the form like this:

![image](https://user-images.githubusercontent.com/12464858/92182153-5d028880-ee21-11ea-8cd4-b19f69f14647.png)

tiagodread reviewed (changes requested):
There’s some issues that we have to fix first before merging this PR

keithjgrant committed:
convert http headers field to string for codemirror

keithjgrant committed:
fix multiple notification form bugs

keithjgrant commented:
@tiagodread I've addressed all the issues listed above, except the following items:

> Trying to create a notification without organization the message "This field is required." is showing twice

This is a known bug I plan to address in https://github.com/ansible/awx/issues/7515
I did tweak the layout a little so the first error shows in red as it is meant to and the second error sits beneath the last form field.

> After saving a notification on the details tab the metadata `Destination Channels` shows all destinations without any space and this happens on all the places where you can specify multiple channels or emails to receive a notification

I've fixed this bug in a separate forthcoming PR

> Trying to create a notification type twilio the fields Source phone number and Destination SMS number(s) are accepting letters, is this correct ?

This matches behavior of the old UI, though it certainly could be improved in the future (maybe file a low priority issue for follow-up? phone number validation is notoriously difficult to do well, made even more so by the multi-input field here)

> Trying to create a email notification type, the fields Host, Recipient list, Port and Time out doesn’t have validation

I fixed `port` and `timeout` validation, but let the other fields un-validated to match behavior of the old UI.

keithjgrant committed:
handle error string in FormSubmitError

tiagodread commented:
@keithjgrant is it possible to add unique ids for these custom message inputs elements?

keithjgrant committed:
add ids to CodeMirror FormGroup containers

tiagodread reviewed (approved):
Retested, tests and stubs are written!
👍🏽
